FFT for OMP backend (via 2decomp&fft)

Nanoseb opened the pull request:
closes #54

Nanoseb commented:
@CFD-Xing I feel like you added a lot of unnecessary files in the last commit. Can you remove them?
`CMakeCache.txt` shouldn't be in the repository, same as `CMakeDetermineCompilerABI_Fortran.bin` or anything inside `cmake/CMakeFiles/3.28.3` or `cmake/CMakeFiles/CMakeConfigureLog.yaml` and maybe more.

The build worked before, you just needed to build 2decomp&fft and then setup an environment variable named `decomp2d_INCLUDE_DIRS` or something like that where the installation folder was and it was picked up fine by x3d2. We also want to make sure 2decomp is still an optional dependency of x3d2 which doesn't seem to be the case anymore.

semi-h commented:
@CFD-Xing How does it work now compared to what we have in the ADIOS2 branch?

https://github.com/xcompact3d/x3d2/blob/jq/implement-io/.gitmodules
https://github.com/xcompact3d/x3d2/blob/jq/implement-io/CMakeLists.txt

pbartholomew08 commented on src/CMakeLists.txt:
```diff
@@ -78,7 +78,6 @@ if (${POISSON_SOLVER} STREQUAL "FFT" AND ${BACKEND} STREQUAL "OMP")
   message(STATUS "Using the FFT poisson solver with 2decomp&fft")
   set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake")
   find_package(decomp2d REQUIRED)
-  include_directories(${decomp2d_INCLUDE_DIRS})
```
This line is required to set the include paths for the 2decomp modules when building x3d2

pbartholomew08 commented on cmake/Finddecomp2d.cmake:
```diff
@@ -0,0 +1,27 @@
+else(decomp2d_FOUND)
+  message(STATUS "2decomp-fft PATH not available we'll try to download and install")
+  configure_file(${CMAKE_SOURCE_DIR}/cmake/decomp2d/downloadBuild2decomp.cmake.in decomp2d-build/CMakeLists.txt)
+  #message("Second CMAKE_GENERATOR ${CMAKE_GENERATOR}") 
```
Remove commented code

Nanoseb replied:
done

pbartholomew08 commented on cmake/decomp2d/downloadBuild2decomp.cmake.in:
```diff
@@ -0,0 +1,26 @@
+    TEST_COMMAND      ""
+)
+
+#ExternalProject_Add(downloadBuild2decomp
+#    GIT_REPOSITORY    "https://github.com/xcompact3d/2decomp-fft"
+#    GIT_TAG           "main"
+#    CONFIGURE_COMMAND "cmake -S ${CMAKE_CURRENT_BINARY_DIR}/decomp2d-src "
+#    BUILD_COMMAND     ""
+#    INSTALL_COMMAND   ""
+#    TEST_COMMAND      ""
+#    SOURCE_DIR        "${CMAKE_CURRENT_BINARY_DIR}/decomp2d-src"
+#    BINARY_DIR        ""
+#    INSTALL_DIR       ""
```
Remove commented code?

pbartholomew08 commented on src/omp/backend.f90:
```diff
@@ -418,7 +418,9 @@ subroutine vecadd_omp(self, a, x, b, y)
     end if
 
-    dims = size(x%data)
+    !dims = size(x%data)
+    ! Fix for size being stored wrongly into dims
+    dims = self%mesh%get_padded_dims(x)
```
Remove comment and/or replace with explanation (to prevent people reverting back if non-obvious)

Nanoseb replied:
done

pbartholomew08 commented on src/xcompact.f90:
```diff
@@ -22,9 +22,9 @@ program xcompact
-  type(mesh_t) :: mesh
   type(allocator_t), pointer :: host_allocator
   type(solver_t) :: solver
+  type(mesh_t), target :: mesh
```
Should this be `class` based on `src/allocator.f90`?

Nanoseb replied:
`type` was correct, allocator has been fixed.

pbartholomew08 commented on tests/omp/test_omp_transeq.f90:
```diff
@@ -47,7 +47,7 @@ program test_omp_transeq
 
   ! Domain decomposition in each direction
-  nproc_dir = [nproc, 1, 1]
+  nproc_dir = [1, 1, nproc]
```
Any particular reason for this change? Should we consider running variations?

Nanoseb replied:
Yes, this is because 2decomp&fft only allows 2D decomposition and as it is implemented at the moment it is in y and z. I think that's also a limitation of the cuda backend, but I am not sure anymore. 
That could likely be changed in the future though, whether this should be part of this PR or not I am not sure.

pbartholomew08 commented on tests/test_reordering.f90:
```diff
@@ -75,7 +75,7 @@ program test_reorder
 
   ! Domain decomposition in each direction
-  nproc_dir = [nproc, 1, 1]
+  nproc_dir = [1, 1, nproc]
```
As for `test_omp_transeq.f90`

Nanoseb replied:
Yes, this is because 2decomp&fft only allows 2D decomposition and as it is implemented at the moment it is in y and z.

pbartholomew08 commented on src/2decompfft/decomp.f90:
```diff
@@ -0,0 +1,113 @@
+    grid%cell_dims(:) = xsize(:)
+    par%n_offset(:) = xstart(:)
+
+    ! compute vert_dims from cell_dims
+    do dir = 1, 3
+      is_last_domain = (par%nrank_dir(dir) + 1 == par%nproc_dir(dir))
+      if (is_last_domain .and. (.not. grid%periodic_BC(dir))) then
+        grid%vert_dims(dir) = grid%cell_dims(dir) +1
+      else
+        grid%vert_dims(dir) = grid%cell_dims(dir)
+      end if
+    end do
+
+    ! Get neighbour ranks
+    do dir = 1, 3
+      nproc = par%nproc_dir(dir)
+      subd_pos_prev(:) = subd_pos(:)
+      subd_pos_prev(dir) = modulo(subd_pos(dir) - 2, nproc) + 1
+      par%pprev(dir) = global_ranks(subd_pos_prev(1), &
+                                         subd_pos_prev(2), &
+                                         subd_pos_prev(3))
+
+      subd_pos_next(:) = subd_pos(:)
+      subd_pos_next(dir) = modulo(subd_pos(dir) - nproc, nproc) + 1
+      par%pnext(dir) = global_ranks(subd_pos_next(1), &
+                                         subd_pos_next(2), &
+                                         subd_pos_next(3))
+    end do
```
Is this not handled by mesh initialisation? i.e. is this the appropriate place to do this, or should it be done by a `mesh_2decomp_t` object specialisation?

Nanoseb replied:
This has been refactored into and isn't done in the decomp code anymore

pbartholomew08 commented on src/CMakeLists.txt:
```diff
@@ -31,21 +36,31 @@ set(CUDASRC
+if (${POISSON_SOLVER} STREQUAL "FFT" AND ${BACKEND} STREQUAL "OMP")
+  list(APPEND SRC ${2DECOMPFFTSRC})
+else()
+  list(APPEND SRC ${GENERICDECOMPSRC})
```
Does `GENERIC` do nothing? Will this be an issue for CUDA code?

Nanoseb replied:
Generic does the decomposition fine, and should be used when using CUDA (or ITER). It may not be consistent with what 2decomp&fft does internally hence the need to have a separate one for it calling 2decomp directly.

pbartholomew08 commented on src/allocator.f90:
```diff
@@ -64,7 +64,7 @@ module m_allocator
 
   function allocator_init(mesh, sz) result(allocator)
-    type(mesh_t), target, intent(inout) :: mesh
+    class(mesh_t), target, intent(inout) :: mesh
```
Do we have mesh subtypes now?

Nanoseb replied:
changed back to type

pbartholomew08 commented on src/decomp.f90:
```diff
@@ -0,0 +1,54 @@
+module m_decomp
```
Module could use some comments

pbartholomew08 commented on src/omp/kernels/spectral_processing.f90:
```diff
@@ -0,0 +1,99 @@
+      do j=1, ny_spec
+        do i=1, nx_spec 
+          ! normalisation
+          div_r = real(div_u(i, j, k), kind=dp)/(nx*ny*nz)
+          div_c = aimag(div_u(i, j, k))/(nx*ny*nz)
```
In case the mesh is large, either
```
num / nx / ny / nz
```
or
```
num / (int(nx, int64) * int(ny, int64) * int(nz, int64)
```
to avoid overflows

Nanoseb replied:
good point, the CUDA implementation might need the same change as I took most of this from there.

Nanoseb replied:
done for both the cuda and omp backends

pbartholomew08 commented on src/poisson_fft.f90:
```diff
@@ -21,7 +21,7 @@ module m_poisson_fft
     complex(dp), allocatable, dimension(:, :, :) :: waves
     !> Wave numbers in x, y, and z
-    complex(dp), allocatable, dimension(:) :: ax, bx, ay, by, az, bz
+    real(dp), allocatable, dimension(:) :: ax, bx, ay, by, az, bz
```
Is this safe for cuFFT?

semi-h reviewed (commented):
It looks like when x3d2 is set up with 2DECOMP&FFT support the decomposition depends on 2DECOMP&FFT and can't be changed at runtime. When the iterative solver is in the mainline we'll have cases where a Cartesian decompostion would't work, and currenly the only option to run such a case would be recompiling without 2DECOMP&FFT. There would be a similar issue when CUDA backend supports 2DECOMP&FFT as well. 

It would be better to have an option like -DWITH_2DECOMP and then if present compiling and making it possible to use 2DECOMP&FFT optionally with runtime parameters. My suggestion below would be a solution I think.

Nanoseb commented on src/mesh.f90:
```diff
@@ -7,43 +7,17 @@ module m_mesh
     integer, private :: sz
     type(geo_t), allocatable :: geo ! object containing geometry information
-    type(parallel_t), allocatable :: par ! object containing parallel domain decomposition information
+    class(grid_t), allocatable :: grid ! object containing grid information
```
I agree with defining these inside mesh.f90, but that isn't possible due to limitations of the compilers. They flag circular dependencies when doing so (even if there aren't any) just because they compile mesh.f90 at once and can't compile the modules independently. This is the primary reason why all the contents of the mesh object have to be in a separate module (for actual circular dependency) and in a separate file (compiler limitation).

CFD-Xing commented:
@Nanoseb 

```
I think I have a strategy to force the re-build of decomp2d-fft, if desired. Here is the cmake

# - Find the 2decomp-fft library
if (rebuild_decom2d)
    message("Re-build 2decomp-fft")
    execute_process(COMMAND rm -rf ${CMAKE_CURRENT_BINARY_DIR}/decomp2d-build)
    set(decomp2d_FOUND FALSE)
else()
    find_package(decomp2d
                 PATHS ${CMAKE_SOURCE_DIR}/decomp2d/build)
endif()

if (decomp2d_FOUND)
  message(STATUS "2decomp-fft FOUND")
else(decomp2d_FOUND)
  message(STATUS "2decomp-fft PATH not available we'll try to download and install")
  configure_file(${CMAKE_SOURCE_DIR}/cmake/decomp2d/downloadBuild2decomp.cmake.in decomp2d-build/CMakeLists.txt)
  execute_process(COMMAND ${CMAKE_COMMAND} -G "${CMAKE_GENERATOR}" .
          RESULT_VARIABLE result
          WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/decomp2d-build )
  if(result)
      message(FATAL_ERROR "CMake step for 2decomp-fft failed: ${result}")
  else()
      message("CMake step for 2decomp-fft completed (${result}).")
  endif()
  execute_process(COMMAND ${CMAKE_COMMAND} --build .
         RESULT_VARIABLE result
          WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/decomp2d-build )
  if(result)
      message(FATAL_ERROR "Build step for 2decomp-fft failed: ${result}")
  endif()
  set(D2D_ROOT ${CMAKE_CURRENT_BINARY_DIR}/decomp2d-build/downloadBuild2decomp-prefix/src/downloadBuild2decomp-build)
  find_package(decomp2d REQUIRED
          PATHS ${D2D_ROOT})
endif(decomp2d_FOUND)
```

You can turn on/off the option by using 

`cmake .. -DCMAKE_BUILD_TYPE=Debug -Drebuild_decom2d=ON`

`cmake .. -DCMAKE_BUILD_TYPE=Debug -Drebuild_decom2d=OFF`

Feel free to include this in your PR if you are happy with the behaviour.

Nanoseb commented on src/2decompfft/decomp.f90:
```diff
@@ -0,0 +1,113 @@
+    ny = grid%global_cell_dims(2)
+    nz = grid%global_cell_dims(3)
+
+    p_row = par%nproc_dir(2)
+    p_col = par%nproc_dir(3)
+    if (p_row*p_col /= par%nproc) then
+      error stop "Decomposition in X not supported by 2decomp&fft backend"
+    end if
+    periodic_bc(:) = grid%periodic_BC(:)
+    call decomp_2d_init(nx, ny, nz, p_row, p_col, periodic_bc)
+
+    ! Get global_ranks
+    allocate(global_ranks(1, p_row, p_col))
+    allocate(global_ranks_lin(p_row*p_col))
+    global_ranks_lin(:) = 0
+
+    call MPI_Comm_rank(DECOMP_2D_COMM_CART_X, cart_rank, ierr)
+    call MPI_Cart_coords(DECOMP_2D_COMM_CART_X, cart_rank, 2, coords, ierr)
+
+    global_ranks_lin(coords(1)+1 + p_row*(coords(2))) = par%nrank
+
+    call MPI_Allreduce(MPI_IN_PLACE, global_ranks_lin, p_row*p_col, MPI_INTEGER, MPI_SUM, MPI_COMM_WORLD, ierr)
+
+    global_ranks = reshape(global_ranks_lin, shape=[1, p_row, p_col])
```
I think the main issue is that 2decomp doesn't only gives us the rank mapping, it also sets the number of cells in each direction per rank and the cell offset. These two things are more mesh related which makes this function a bit more tied in with the mesh object and initialisation. I am still thinking about the best way to do this and in particular reducing the code duplication between the generic and 2decomp decompositions.

semi-h replied:
I think we can pass cell numbers as well as rank mapping when instantantiating the mesh object. The offset can then be determined easily from the cell numbers during the mesh object initialisation. Is there anything else we ask 2decomp to provide?

What I have in mind is initialising 2decomp somewhere in the main program just before mesh. Then we have the outputs from 2decomp: rank mapping and local cell sizes. We can pass these two as arguments when instantiating the mesh object. We can have these as optional arguments so we fall back to default trivial mappings in tests. Also, we can do this 2decomp initialision with ifdefs, so it is called only when x3d2 is compiled with 2decomp support. If mesh object is not given optional domain mapping info and cell count info we use our current strategy. (or when domain mapping and cell data is all zeros, meaning no preference just to make it easier)

Nanoseb replied:
yeah, sure. That's doable. I will try to avoid ifdef though, the current implementation was designed to not have any ifdef and only rely on the linking to select which implementation to choose from. I feel like that's cleaner and allows us to add more decomposition methods later on without any change to the current code.
This means that, even without 2decomp, the generic decomposition will still be done prior to the mesh object instantiation etc.

Overall, I am not sure I fully see the benefits of having the decomposition outside the mesh init as the two are intertwined to me, but that's fine, I will do that.

semi-h replied:
I think the benefit is that we can remove `decomp.f90`, `2decompfft/decomp.f90`, and `decomp_generic.f90` files, and this new structure around `m_decomp` modules if we allow mesh to take a 'suggestion' regarding decomposition. We can still have any custom decomposition and can implement this inside the mesh.

Not having ifdef is better definitely, can you comment on whether this would allow us to change between FFT and iterative solver, and potentially changing between using 2decomp or a different library (more related to CUDA backend, but better to not prevent this with the structure we design here).

semi-h replied:
I mean changing between FFT/iteretive solver at runtime. For example its likekly we'll have a case with stretching in two/three directions, and then FFT wouldn't be compatible with that case, and we'll have to use iterative with it. Ideally we shouldn't make a compile time decision between these solvers and should be select at runtime.

slaizet replied:
We might want to have a discussion about how to handle two different Poisson solvers (at compilation or at runtime or something else), in terms of user friendliness, performance, and ease of implementation :-)

Nanoseb replied:
Yeah, having a chat about it is probably a good idea, also because @pbartholomew08 has implemented something similar for his iterative poisson solver. It would be good to coordinate to have a consistent solution for these things.

pbartholomew08 replied:
I can build my branch without PETSc, it requires a dummy module to link against which will simply crash the program if you try to use iterative solvers (untested). And it should be possible to switch at runtime between FFT/CG (untested).

However, this may be relatively easier to do than with a low level data structure as in x3d2-mesh vs 2decomp - perhaps if we want that kind of flexibility we need to accept data copies between "x3d2 space" and "2decomp space", this is currently what I have in the PETSc module too, but I hope to eliminate.

The dummy module is simply to enable compilation:
```
submodule(m_poisson_cg) m_dummy_poisson_cg
  use m_base_backend, only: base_backend_t
contains
  module subroutine init_precon_impl(precon, backend)
    class(poisson_precon_impl_t), allocatable, intent(out) :: precon
    class(base_backend_t), intent(in) :: backend
    error stop "This build doesn't support iterative Poisson solvers"
  end subroutine
  module subroutine apply_precon_internal(self, p, b, backend)
    class(poisson_precon_impl_t) :: self
    class(field_t), intent(in) :: p ! Pressure solution
    class(field_t), intent(inout) :: b ! The evaluated matrix-vector product
    class(base_backend_t), intent(in) :: backend
    error stop "This build doesn't support iterative Poisson solvers"
  end subroutine apply_precon_internal
  module subroutine init_solver(solver, backend)
    class(poisson_solver_t), allocatable, intent(out) :: solver
    class(base_backend_t), target, intent(in) :: backend
    error stop "This build doesn't support iterative Poisson solvers"
  end subroutine init_solver
  module subroutine solve_poisson(self, p, f, backend)
    class(poisson_solver_t) :: self
    class(field_t), intent(inout) :: p ! Pressure solution
    class(field_t), intent(in) :: f    ! Poisson RHS
    class(base_backend_t), intent(in) :: backend
    error stop "This build doesn't support iterative Poisson solvers"
  end subroutine solve_poisson
end submodule
```

Nanoseb replied:
I think the main difference currently between my implementation and Paul's one is that I don't have a dummy module, instead, I have the 'generic' decomposition implementation that is being used when 2decomp isn't available. This doesn't allow runtime switch.

So what I am going to do now is have that dummy module like Paul and move the generic decomposition somewhere else that is always compiled. That way we can choose at run time which one we use while still avoiding ifdef etc.
For now I will leave this inside the mesh initialisation, but it can easily be moved outside of it later on if needed.

Nanoseb replied:
It's been decided that to stick with a submodule method for now and we might move to a different method in a future PR.

semi-h reviewed (commented):
Commented mainly on some minor stuff.

Before I forget to mention, because a few different approaches have been tried here and added/removed plenty of stuff, is there a way to keep the history of the main branch clean of all these things? Should help people browsing the history but not sure what is the best approach. Maybe squash and merge option or the rebase one when merging?

pbartholomew08 commented:
> Commented mainly on some minor stuff.
> 
> Before I forget to mention, because a few different approaches have been tried here and added/removed plenty of stuff, is there a way to keep the history of the main branch clean of all these things? Should help people browsing the history but not sure what is the best approach. Maybe squash and merge option or the rebase one when merging?

We could squash/merge or interactively rebase a branch before merging it. Interactive rebase before merge is perhaps better as you can be selective about what happens, squash just turns the entire branch into a single commit (which may be undesirable).

Nanoseb commented on src/2decompfft/omp/poisson_fft.f90:
Done

Nanoseb commented on tests/test_fft.f90:
```diff
@@ -0,0 +1,161 @@
+program test_reorder
```
fixed

Nanoseb commented on src/CMakeLists.txt:
```diff
@@ -59,11 +74,19 @@ elseif(${CMAKE_Fortran_COMPILER_ID} STREQUAL "GNU")
 endif()
 
+if (${POISSON_SOLVER} STREQUAL "FFT" AND ${BACKEND} STREQUAL "OMP")
+  message(STATUS "Using the FFT poisson solver with 2decomp&fft")
```
Done, now it is defined only based on the fact that 2decomp is available and fft is selected.

Nanoseb commented on src/CMakeLists.txt:
```diff
@@ -31,21 +36,31 @@ set(CUDASRC
   list(APPEND SRC ${CUDASRC})
 endif()
 
+if (${POISSON_SOLVER} STREQUAL "FFT" AND ${BACKEND} STREQUAL "OMP")
```
I called it `2DECOMPFFT_SUPPORT` but `WITH_2DECOMPFFT` is probably better

Nanoseb replied:
Just renamed the variable

Nanoseb commented on src/xcompact.f90:
```diff
@@ -43,9 +43,13 @@ program xcompact
+  logical :: use_2decomp
 
   namelist /domain_params/ L_global, dims_global, nproc_dir, BC_x, BC_y, BC_z
+  namelist /solver_params/ poisson_solver_type
```
What compiler was it with? it seems to be working fine with gnu. Could you double check before I move it to a new namelist?

Regardless of this, I think longer term it would be good to read the input file only once and make it populate an `option` or `parameter` object that will have everything. That way it can be accessed easily by the codebase without re-reading the file

Nanoseb replied:
ah no, my bad, it also doesn't work with GNU, I will fix that.

Nanoseb replied:
Just done this by creating a reader for the solver input. Seemed cleaner than alternatives. It will probably be good to do something similar for the domain parameters but it is a bit out of the scope of this PR.

Nanoseb commented:
I think it is ready to be merged. I have addressed all the comments and left 'unresolved' the ones that may need double checking by one of you (i.e. I have resolved the trivial ones but left unresolved the others).

I am tempted to have the cmake auto build of 2decomp in a separate branch because it likely requires a discussion to have single solution for all the dependencies (petsc etc.). Plus it doesn't allow to select an other 2decomp backend (like fftw).

semi-h commented:
> I think it is ready to be merged. I have addressed all the comments and left 'unresolved' the ones that may need double checking by one of you (i.e. I have resolved the trivial ones but left unresolved the others).

I'll quickly check if all good with the CUDA backend and fix any minor issues if there is any, I think we can aim to merge by the end of today.

pbartholomew08 commented on src/CMakeLists.txt:
```diff
@@ -32,39 +37,61 @@ set(CUDASRC
 
   target_compile_options(xcompact PRIVATE "-DCUDA")
-elseif(${CMAKE_Fortran_COMPILER_ID} STREQUAL "GNU")
+elseif(${COMPILER} STREQUAL "GNU")
```
Why this change? It's probably better to stick to the CMAKE variables

pbartholomew08 replied:
if this is a pain to fix we could lift it into a separate issue.

Nanoseb replied:
The "COMPILER" is directly based on the CMAKE variable, I found it cleaner to be consistent with cuda. But happy to switch back, it turns out it is used only in this if statement, so making an extra variable isn't as needed as it used to.

fyi, the main cmakelists.txt has:

```cmake
if(${CMAKE_Fortran_COMPILER_ID} STREQUAL "PGI" OR
   ${CMAKE_Fortran_COMPILER_ID} STREQUAL "NVHPC")
   set(COMPILER CUDA)
elseif(${CMAKE_Fortran_COMPILER_ID} STREQUAL "GNU")
   set(COMPILER GNU)
else()
   set(COMPILER OTHER)
endif()
```

pbartholomew08 replied:
ok, then we leave this to be consistent with the rest of the code.

Nanoseb replied:
Just changed it back, I think you are right, it is better to not use 'custom' variables when it isn't really needed. And because it is used only here there is no point in having it. It was more useful before when the compiler and backend were the same thing, but because that isn't the case anymore, it isn't as needed.

pbartholomew08 commented on tests/test_fft.f90:
```diff
@@ -0,0 +1,164 @@
+  call backend%poisson_fft%fft_backward(output_field)
+
+  ! Output scaled with number of cells in domain
+  error_norm = norm2(input_field%data(:, :, :) - output_field%data(:, :, :)/product(dims_global) )**2/product(dims_global)
```
Just to double check - the first division by problem size is normalising the backwards FFT, i.e. intentional, and 2nd is part of RMS operation?

Nanoseb replied:
yes exactly. Do you think it needs extra comments?

Nanoseb replied:
you are right, just added one.

pbartholomew08 reviewed (approved):
All looks good to me

semi-h commented:
Had some trouble compiling with `nvfortran`. First got this module prefix error;
```
NVFORTRAN-S-1056-MODULE prefix is only allowed for subprograms that were declared as separate module procedures (x3d2/src/decomp/decomp.f90: 49)
  0 inform,   0 warnings,   1 severes, 0 fatal for decomp_grid
NVFORTRAN-S-1056-MODULE prefix is only allowed for subprograms that were declared as separate module procedures (x3d2/src/decomp/decomp.f90: 57)
```
Not sure if its safe, but was able to remote module prefixes in subroutines/functions and compile fine.
Then now having a trouble with
```
/usr/bin/ld: libx3d2.a(decomp.f90.o): in function `m_decomp_init_decomp_mod_':
x3d2/src/decomp/decomp.f90:45:(.text+0x89): undefined reference to `m_decomp_init_decomp_'
```

Not sure yet what is causing this, but it smells bad
https://fortran-lang.discourse.group/t/support-for-submodules-in-nvfortran/5097

Maybe nvfortran doesn't yet support submodules, not sure the full extent, still trying to figure out but wanted to share where I'm at.

Nanoseb commented:
arg, ffs. Why can't they follow the standards!

Nanoseb commented:
This last commit removes submodule because of lack of support from nvidia. I've done it by defining 2 modules (dummy and actual one) with the same functions. The switch between the two is done by cmake just like before. The main issue with this approach is that we have to keep the consistency of the two module ourselves, we can't rely on the submodule structure to enforce that. Though failing to do so will lead to a compiler error so that can be detected early on.

@pbartholomew08 I'd be interested to hear if you have comments about this approach.

semi-h commented on src/decomp/decomp_2decompfft.f90:
```diff
@@ -0,0 +1,74 @@
+
+    ! Get local domain size and offset from 2decomp
+    grid%cell_dims(:) = xsize(:)
+    par%n_offset(:) = xstart(:)
```
Do we need a -1 here as well?

Nanoseb replied:
It is fine because it is used the same way in 2decomp and x3d2. We might want to remove the -1 from x/yz_sp_st definition and add it when it is used actually, for consistency.

Nanoseb commented on src/decomp/decomp_2decompfft.f90:
```diff
@@ -64,7 +64,7 @@ subroutine decomposition_2decomp(grid, par)
     ! Get local domain size and offset from 2decomp
     grid%cell_dims(:) = xsize(:)
-    par%n_offset(:) = xstart(:)
+    par%n_offset(:) = xstart(:) - 1
```
Are you sure about that one? if this is changed then https://github.com/xcompact3d/x3d2/blob/6be7442c2e30ac7e5f15e2ed96d30d1cf6cf9b34/src/mesh.f90#L470 should also be changed I think

semi-h replied:
Was just about writing a comment on this. Yes, I printed the values to make sure and `xstart` from 2decomp returns the global starting index of the array. What we do in coordinates is making sure that (cartesian field(i, j, k) entry @(i=1, j=1, k=1) correspond to the coordinates (0, 0, 0). I think this is the standard in Xcompact3d too. If we use the starting index instead then we would have to substract another 1 to make sure the coordinate corresponding to (i=1, j=1, k=1) is (0, 0, 0). Because TGV is all periodic at the moment we can't really observe a difference and works fine either way, but would be a problem for example in channel case because the initial conditions depend on the global coordinates.
